File: src/entities/experience.ts
import { experienceOrbs, player, state } from "../state";
import { collectExperience } from "../game/progression";
import { spark } from "./particles";
import {
  experienceDropTotal,
  experienceOrbRadius,
  experienceShardCount,
} from "../game/balance";
import type { EnemyEntity } from "../types";
import { pickupRadiusFor, shouldCollectOrb } from "./experience-pickup";
import { markHudDirty } from "../simulation/events";
import { experienceGrid } from "../simulation/grids";
import { acquireExperienceOrb, releaseExperienceOrb } from "../simulation/pools";
import { random } from "../simulation/random";

const GRID_PICKUP_THRESHOLD = 64;
const pickupCandidateIds = new Set<number>();

export function spawnExperience(enemy: EnemyEntity): void {
  const total = experienceDropTotal(enemy.score, state.wave);
  const shardCount = experienceShardCount(enemy.kind);
  let remaining = total;

  for (let i = 0; i < shardCount; i += 1) {
    const value = i === shardCount - 1 ? remaining : Math.max(1, Math.round(total / shardCount));
    remaining -= value;
    const angle = random() * Math.PI * 2;
    const speed = 70 + random() * 120;
    const orb = acquireExperienceOrb();
    orb.x = enemy.x;
    orb.y = enemy.y;
    orb.vx = Math.cos(angle) * speed;
    orb.vy = Math.sin(angle) * speed;
    orb.radius = experienceOrbRadius(value);
    orb.value = value;
    orb.age = random() * 0.4;
    orb.magnetized = false;
  }
}

export function updateExperience(dt: number): void {
  const damp = 1 - dt * 2.7;
  let hasLooseOrbs = false;

  for (let i = experienceOrbs.length - 1; i >= 0; i -= 1) {
    const orb = experienceOrbs[i]!;
    orb.age += dt;
    orb.x += orb.vx * dt;
    orb.y += orb.vy * dt;
    orb.vx *= damp;
    orb.vy *= damp;
    hasLooseOrbs ||= !orb.magnetized;
  }

  const usePickupGrid = hasLooseOrbs && experienceOrbs.length > GRID_PICKUP_THRESHOLD;
  pickupCandidateIds.clear();
  if (usePickupGrid) {
    experienceGrid.rebuild(experienceOrbs);
    experienceGrid.visitRadius(player.x, player.y, pickupRadiusFor(player), (orb) => {
      if (!orb.magnetized) {
        pickupCandidateIds.add(orb.id);
      }
    });
  }

  for (let i = experienceOrbs.length - 1; i >= 0; i -= 1) {
    const orb = experienceOrbs[i]!;
    if (!orb.magnetized && usePickupGrid && !pickupCandidateIds.has(orb.id)) continue;
    if (shouldCollectOrb(orb, player, dt)) {
      collectExperience(orb.value);
      spark(orb.x, orb.y, "#72ffb1");
      releaseExperienceOrb(i);
      markHudDirty();
    }
  }
}

File: sim/crates/voidline-sim/src/experience.rs
//! Experience orb spawning + collection owned by the Rust runtime engine.

use voidline_data::balance::Balance;

use crate::balance_curves::{experience_drop_total, experience_orb_radius, experience_shard_count};
use crate::entities::{Enemy, ExperienceOrb};
use crate::player::Player;
use crate::pools::{acquire_experience_orb, release_experience_orb, EntityPools};
use crate::progression::collect_experience;
use crate::rng::Mulberry32;
use crate::state::{EntityCounters, GameState};

const MAGNETIZED_PULL: f64 = 560.0;
const MAGNETIZED_CONTACT_PADDING: f64 = 8.0;

pub fn spawn_experience(
    balance: &Balance,
    pools: &mut EntityPools,
    counters: &mut EntityCounters,
    orbs: &mut Vec<ExperienceOrb>,
    rng: &mut Mulberry32,
    enemy: &Enemy,
    pressure: u32,
) {
    let total = experience_drop_total(balance, enemy.score, pressure);
    let shard_count = experience_shard_count(balance, enemy.kind.as_str());
    let mut remaining = total;

    for i in 0..shard_count {
        let value = if i == shard_count - 1 {
            remaining
        } else {
            ((total as f64) / (shard_count as f64)).round().max(1.0) as i64
        };
        remaining -= value;
        let angle = rng.next_f64() * std::f64::consts::TAU;
        let speed = 70.0 + rng.next_f64() * 120.0;
        let radius = experience_orb_radius(balance, value as f64);
        let age = rng.next_f64() * 0.4;
        let idx = acquire_experience_orb(pools, counters, orbs);
        let orb = &mut orbs[idx];
        orb.x = enemy.x;
        orb.y = enemy.y;
        orb.vx = angle.cos() * speed;
        orb.vy = angle.sin() * speed;
        orb.radius = radius;
        orb.value = value as f64;
        orb.age = age;
        orb.magnetized = false;
    }
}

pub fn pickup_radius_for(balance: &Balance, player: &Player) -> f64 {
    balance.xp.pickup_base_radius * player.pickup_radius
}

pub fn should_collect_orb(
    orb: &mut ExperienceOrb,
    player: &Player,
    pickup_radius: f64,
    dt: f64,
) -> bool {
    let dx = player.x - orb.x;
    let dy = player.y - orb.y;
    let dist_sq = dx * dx + dy * dy;
    if orb.magnetized {
        let distance = dist_sq.sqrt();
        let inv = 1.0 / distance.max(1.0);
        orb.vx += dx * inv * MAGNETIZED_PULL * dt;
        orb.vy += dy * inv * MAGNETIZED_PULL * dt;
        let contact = player.radius + orb.radius + MAGNETIZED_CONTACT_PADDING;
        return dist_sq < contact * contact;
    }
    dist_sq < pickup_radius * pickup_radius
}

pub fn update_experience(
    balance: &Balance,
    pools: &mut EntityPools,
    state: &mut GameState,
    player: &mut Player,
    orbs: &mut Vec<ExperienceOrb>,
    dt: f64,
) {
    let damp = 1.0 - dt * 2.7;
    for orb in orbs.iter_mut() {
        orb.age += dt;
        orb.x += orb.vx * dt;
        orb.y += orb.vy * dt;
        orb.vx *= damp;
        orb.vy *= damp;
    }

    let pickup_radius = pickup_radius_for(balance, player);
    let mut i = orbs.len();
    while i > 0 {
        i -= 1;
        let collect = should_collect_orb(&mut orbs[i], player, pickup_radius, dt);
        if collect {
            let value = orbs[i].value as u32;
            collect_experience(balance, state, player, value);
            release_experience_orb(pools, orbs, i);
        }
    }
}
